build: add a cmake target for all used linters

dundargoc opened the pull request:
build: add a cmake target for all used linters

Part 1 of the cmake saga (this one)
Part 2 is going to be checking the uncrustify version is OK, and warn if it isn't.
Part 3 is about creating a `make format` target that format everything.

dundargoc commented:
Would like this merged before working on uncrustify CI as that will have at least 3 more optional commands if my calculations are correct.

justinmk commented:
> All other program checks are done in Makefile except for lualint, which
> is checked in CMakeLists.txt instead. Having the check in the Makefile
> is more consistent

It should probably be the reverse. CMake is our source of truth. Makefile is a thin convenience, it's not supposed to become the source of truth. People should be able to get all build tasks cmake only.

Makefile tasks are not available on Windows, xcode, ninja, etc.

dundargoc commented:
> CMake is our source of truth.

We live in the darkest timeline.

I'll move the function to cmake instead :+1:

dundargoc committed:
build: move the logic for linters to cmake

Cmake is our source of truth. We should have as much of our build
process there as possible so everyone can make us of it.

dundargoc committed:
build: remove redundant check for ninja generator

The minimum cmake version as of writing this is 3.10, which has ninja
support.

justinmk commented on CMakeLists.txt:
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
-    COMMENT "lualint: LUACHECK_PRG not defined")
-endif()
-
+find_program(LUACHECK luacheck)
```
the `_PRG` suffix convention was intentional and used throughout our cmake files. Why revert that?

dundargoc replied:
Found it too wordy, don't mind either way.

justinmk replied:
`_PRG` convention should be kept

justinmk commented on CMakeLists.txt:
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
+  -DTARGET=${TARGET}
+  -P ${PROJECT_SOURCE_DIR}/cmake/lint.cmake)
+
+set(TARGET "stylua")
```
all lint tasks should have `lint` suffix. Or better, a `lint` prefix, since we're changing things already.

justinmk commented on Makefile:
```diff
@@ -47,13 +47,7 @@ endif
-      BUILD_TOOL = $(MAKE)
-      CMAKE_GENERATOR := Unix Makefiles
-    endif
+    BUILD_TOOL = ninja
```
why remove this fallback?

dundargoc replied:
It's dead code. It's only useful if user's version of cmake doesn't support ninja, but our minimum required cmake version is 3.10, which supports ninja.

justinmk replied:
This fallback was for users that don't have `ninja` installed.

dundargoc replied:
I built neovim without ninja installed, it defaults to make (which I forgot how slow it was, dear lord).

justinmk commented on cmake/lint.cmake:
```diff
@@ -0,0 +1,35 @@
+if(PROGRAM)
```
If we're trying to de-clutter the main `CMakeLists.txt`, can we move all of the `add_custom_target` calls into this file?  It doesn't need to be parameterized.

dundargoc replied:
I'd need to create an additional file in that case due to the limitations of `add_custom_target`:
- it can't silence output
- overall very inflexible since it won't allow you to run other commands.

justinmk replied:
ok but this file isn't doing much then. It's just shuffling some text into a switch-case.

dundargoc replied:
https://github.com/uncrustify/uncrustify/pull/3697

dundargoc committed:
fixup! build: move the logic for linters to cmake

dundargoc committed:
fixup! build: move the logic for linters to cmake

justinmk commented on ci/run_lint.sh:
```diff
@@ -12,7 +12,7 @@ rm -f "$END_MARKER"
 # Run all tests if no input argument is given
 if (($# == 0)); then
-  tests=('clint-full' 'lualint' 'pylint' 'shlint' 'check-single-includes')
+  tests=('lint_c_full' 'lint_lua' 'lint_py' 'lint_sh' 'check-single-includes')
```
I don't think underscores are used in any of our other build targets, so let's stay consistent. Could use hyphens I guess, which is in some of our build targets, but that is annoying to type/complete.

```suggestion
  tests=('lintcfull' 'lintlua' 'lintpy' 'lintsh' 'check-single-includes')
```

justinmk commented on CMakeLists.txt:
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
+  -DPROGRAM=${LUACHECK_PRG}
+  -DPROJECT_ROOT=${PROJECT_SOURCE_DIR}
+  -DTARGET=${TARGET}
+  -P ${PROJECT_SOURCE_DIR}/cmake/lint.cmake)
```
cmake is forcing us to be verbose here, nothing we can do about that. But this means `cmake/lint.cmake` gains little. Just put the extra 1-2 lines here, the indirection of `cmake/lint.cmake` isn't justified to save 1-2 lines here.

dundargoc committed:
fixup: don't use hyphens/underscores for lint targets

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

dundargoc committed:
fixup: add whitespace to retrigger CI

justinmk commented on .github/workflows/ci.yml:
```diff
@@ -107,11 +107,11 @@ jobs:
 
       - if: "!cancelled()"
-        name: stylua
+        name: lintstylua
```
why isn't this part of `lintlua` ?

dundargoc replied:
Wait, this becomes super difficult without having an external script, because we'd need four separate add_custom_target for all cases if stylua/luacheck exists/doesn't exist.

justinmk commented on cmake/lint.cmake:
```diff
@@ -0,0 +1,35 @@
+if(PROGRAM)
+
+  if(${TARGET} STREQUAL "lintuncrustify")
```
why isn't this part of `lintc` ? we don't need separate build targets depending on the tool. We need "lint" tasks, not "lint with a specific tool". The tool is an implementation detail.

dundargoc committed:
fixup! build: move the logic for linters to cmake

justinmk reviewed (approved):
This looks like a nice improvement! 

As a potential followup or bonus, it might be a good idea for our CI to use these targets, that ensures they are "alive" and working. (Except for CI jobs where we don't want to build `nvim`.)

justinmk commented on cmake/lint.cmake:
```diff
@@ -0,0 +1,34 @@
+
+if(${TARGET} STREQUAL "lintuncrustify")
+  file(GLOB_RECURSE FILES ${PROJECT_ROOT}/src/nvim/*.[c,h])
+  lint(PROGRAM uncrustify FLAGS -c src/uncrustify.cfg -q --check FILES ${FILES} QUIET)
```
nit:  consistent casing is helpful, even if it's optional in cmake. The function is defined as `LINT`?

justinmk commented on cmake/lint.cmake:
```diff
@@ -0,0 +1,34 @@
+
+  find_program(PROGRAM_EXISTS ${LINT_PROGRAM})
+  if(PROGRAM_EXISTS)
+    execute_process(COMMAND ${LINT_PROGRAM} ${LINT_FLAGS} ${LINT_FILES}
```
Does this always run, even if there are no changes since the last run? If so, that's a regression. We can think about that in a followup PR though.

justinmk commented on CMakeLists.txt:
```diff
@@ -748,15 +745,13 @@ if(BUSTED_LUA_PRG)
 
-if(LUACHECK_PRG)
-  add_custom_target(lualint
-    COMMAND ${LUACHECK_PRG} -q runtime/ scripts/ src/ test/
```
The old target used `COMMAND ${LUACHECK_PRG}` and the new target uses `execute_process()`. I wonder if we still get the benefits of cmake's "dirty workspace detection".

dundargoc committed:
fixup! build: move the logic for linters to cmake

dundargoc commented:
> As a potential followup or bonus, it might be a good idea for our CI to use these targets, that ensures they are "alive" and working.

They are, albeit in a pretty roundabout way. CI lint steps runs `ci/run_lint.sh` which in turn calls `make` which in turn calls `cmake`. Turtles all the way down :)

justinmk commented:
Eek, this doesn't find `luacheck` on my workspace now:

    -- lintlua: luacheck not found. SKIP.

Does it need to look in .deps ?
